packages/gatsby/src/utils/mett.ts

```diff
@@ -0,0 +1,54 @@
+// This is a simple event emitter based on mitt.js
+// It mainly changes the data model to use a Map and Set, rather than a
+// regular object and an array.
+
+type MettHandler<EventName, Payload> = (
+  e: Payload,
+  eventName: EventName
+) => void
+
+interface IMett<EventName, Payload> {
+  on(eventName: EventName, callback: MettHandler<EventName, Payload>): void
+  off(eventName: EventName, callback: MettHandler<EventName, Payload>): void
+  emit(eventName: EventName, e: Payload): void
+}
+
+function mett<EventName = string, Payload = any>(): IMett<EventName, Payload> {
```

Comment:
I don't think this typing works the way you want it to. The way I am reading this is that I can set this as

```ts
met = mett<'foo', 'bar'>();

met.on('baz', f); // this would fail typescript 
```

Reply:
These sort of APIs are always hard to type. We would have to change the API to get real typing benefits. But that would break the compatibility with mitt.

I think what you want though is something like

```ts
on<EventName, Payload>(eventName: EventName, callback: MettHandler<EventName, Payload>)
```

Then you could describe the specific shape for it.

Reply:
I think there is interface overloading though...

So untested, but this might work to just do something like this:

```ts
// utils/mett.ts

interface IMett {
  on(eventName: '*', callback: MettHandler<'*', any>): void
  off(eventName: '*', callback: MettHandler<'*', any>): void
  emit(eventName: string, e: any): void
}
```

Then in another file something like this:

```
// other-file.ts
interface IMett {
  on(eventName: 'foo', callback: MettHandler<'foo', 'bar'>): void
  off(eventName: 'foo', callback: MettHandler<'foo', 'bar'>): void
  emit(eventName: 'foo', callback: 'bar'): void
}

mett().on('foo', f).emit('foo', 'baz') // should throw error on baz not being bar
```

Reply:
Hm I think I understand what you're saying. By having the type parameter on the whole instance, you basically limit it to one event.

So I guess it's not feasible to make this generic without being specifc? You'd have to be explicit in all the "event name <> handler type" pairs? That makes sense but is way outside the scope of this PR. And feels like quite a big task in general. Not to speak of a bit of a burden for future use (and at the same time a better API documentation, so, :shrug:).

I guess, for this PR, I'll drop back to a simplistic type, using `any` for the data type, and not using type parameters at all.